text/0507-subnets.md

```diff
@@ -0,0 +1,402 @@
+to other hosts inside the VPC. To define different routing policies for a
+subnet, provide a route table when creating it. For example, to create a
+public subnet:
+
+```ts
+const publicRouteTable = vpc.addRouteTable('routeTable', {
+  routes: [
+    // With no parameters, the destination is the entire VPC CIDR
+    Route.local(),
+
+    // By adding this route, all subnets that use this table become public
+    Route.to({
+      destination: '0.0.0.0/0',
+      target: Routers.INTERNET_GATEWAY,
+    }),
+  ],
+});
+
+const subnet = vpc.addSubnet('publicSubnet', {
+  cidrBlock: '10.2.0.0/20',
+  availabilityZone: 'us-west-2a',
+  routeTable: publicRouteTable,
+  mapPublicIpOnLaunch: false, // default: true for public subnets
+});
+
+// The following is true
+vpc.publicSubnets.includes(subnet);
+
+// As is
+vpc.selectSubnets({subnetType: ec2.SubnetType.PUBLIC}).includes(subnet);
+```
+
+To create a route table that sends DynamoDB and S3 traffic through VPC
+endpoints:
+
+```ts
+const routeTable = new RouteTable(this, 'routeTable', {
+  routes: [
+    Route.DYNAMODB_ENDPOINT,
+    Route.S3_ENDPOINT
+  ],
+});
+```
+
+You can use a public NAT gateway to enable instances in a private subnet to
+send outbound traffic to the internet, while preventing the internet from
+establishing connections to the instances:
+
+```ts
+const elasticIp = new ElasticIp({
+  domain: Domain.VPC
+
+  // Other properties, such as networkBorderGroup and publicIpv4Pool, 
+  // are also available. Omitted here for brevity.
+});
+
+const natGateway = publicSubnet.addNatGateway('natgw', {elasticIp});
+
+const routeTable = vpc.addRouteTable('routeTable', {
+  routes: [
+    Route.to({
+      destination: '0.0.0.0/0',
+
+      // targets must implement the IRouter interface
+      target: natGateway,
+    }),
+  ],
+});
+
+const privateSubnet = vpc.addSubnet('privateSubnet', {
+  cidrBlock: '10.2.0.0/20',
+  availabilityZone: 'us-west-2a',
+  routeTable,
+});
+```
+
+You can also create the same kind of routing pattern, but with a NAT instance
+instead:
+
+```ts
+const natInstance = publicSubnet.addNatInstance('natinst', {
+  instanceType: ec2.InstanceType.of(ec2.InstanceClass.BURSTABLE2, ec2.InstanceSize.MICRO),
+  machineImage: ec2.MachineImage.latestAmazonLinux({
+    generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
+  }),
+
+  // Other properties ommitted
+});
+
+const routeTable = vpc.addRouteTable('routeTable', {
+  routes: [
+    Route.to({
+      destination: '0.0.0.0/0',
+
+      // Instance also implements IRouter
+      target: natInstance,
+    }),
+  ],
+});
+
+const privateSubnet = vpc.addSubnet('privateSubnet', {
+  cidrBlock: '10.2.0.0/20',
+  availabilityZone: 'us-west-2a',
+  routeTable,
+});
+```
+
+For IPv6 traffic, you create this pattern by using an egress-only internet
+gateway:
+
+```ts
+const routeTable = vpc.addRouteTable('routeTable', {
+  routes: [
+    Route.local('10.0.0.0/16'),
+    Route.local('2001:db8:1234:1a00:/64'),
```

Comment:
Is this worth autoderiving or is that a problem?

Reply:
The parameter will be optional, with the default value being the entire VPC CIDR. But we do need to have a parameter in case you want to route part of the traffic to the local VPC and part to a peer VPC, for example.